doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,917 @@
+that all "data libraries" can interface with. Persist currently collects
+statistics for filter-pushdown using Arrow.
+
+Parquet is columnar format designed for efficient storage based off of the
+[Dremel paper from Google](https://research.google/pubs/dremel-interactive-analysis-of-web-scale-datasets-2/)
+and built as a [collaboration between Twitter and Cloudera for Hadoop](https://blog.twitter.com/engineering/en_us/a/2013/announcing-parquet-10-columnar-storage-for-hadoop).
+It's goal is space-efficient storage of data that allows fetching individual
+columns from a colletion of rows. Crucially it is also self describing and
+supports additional arbitrary metadata.
+
+Parquet also has a number of nice features that we won't use immediately but
+could enable fairly easily:
+
+* Tracks the size and location of columns within the Parquet file, designed for
+  projection pushdown.
+* Can collect statistics on the data in the Parquet file, which would
+  supplement our own filter pushdown.
+* Can store arbitrary metadata for a file. This could be a way we spill our own
+  collected stats to S3.
+* Dictionary encoding out of the box. For columns with repeated values (e.g.
+  time or diff) could result in space savings.
+* Compression. Individual columns can be compressed with Snappy, gzip, Brotli,
+  or LZ4.
+
+> **Note:** Arrow and Parquet are pretty tightly coupled, mapping from one format to
+the other is already handled by any open-source library we would use. For the
+most part our codebase will interact with Arrow, only the lowest layer of
+Persist having to know about Parquet.
+
+### Columnar Encodings
+
+Currently we durably persist `Datum`s using the [`ProtoDatum`](https://github.com/MaterializeInc/materialize/blob/d0aa5b7d0b47e55cf4e211e507116a41cb7f8680/src/repr/src/row.proto#L32-L69)
+protobuf message. Arrow's data types are not as rich as protobufs, so we need
+to figure out how exactly we would represent `Datum`s in an Arrow format. These
+decisions are not set in stone though, we describe below how migrations to
+other Arrow data types would be possible, but it's less work if we get this
+right from the start.
+
+<table>
+<tr>
+<td> Scalar Type </td><td> Rust Representation </td><td> Arrow Array </td><td> Notes </td>
+</tr>
+
+<tr>
+<td> Numeric </td>
+<td>
+
+```rust
+struct Decimal<const N: usize> {
+  digits: u32,
+  exponent: i32,
+  bits: u8,
+  lsu: [u16; N],
+}
+```
+
+</td>
+<td>
+
+```
+StructArray<{
+  lossy: f64,
+  actual: Bytes,
+}>
+```
+
+</td>
+<td>
+
+Encoded as two values, a lossy `f64` for sorting and filtering, then a
+serialized representation of `Decimal` as opaque bytes.
+
+See <https://speleotrove.com/decimal/dnnumb.html> for an explanation as to what
+the fields in the Rust type represent.
+
+</td>
+
+<tr>
+<td> Date </td>
+<td>
+
+```rust
+struct Date {
+  days: i32,
+}
+```
+
+</td>
+<td>
+
+`PrimitiveArray<i32>`
+
+</td>
+<td>
+
+Directly encode the number of days.
```

Comment:
since the Unix epoch?

Reply:
Today `Date` in Rust is since the Postgres epoch, but `Timestamp` is since the Unix epoch. Updated with the opinion that we should encode as number of days since the Unix epoch for consistency.